indico/modules/receipts/controllers.py

```diff
@@ -0,0 +1,409 @@
+            raise Forbidden
+
+
+class ReceiptTemplateMixin(ReceiptAreaMixin):
+    def _process_args(self):
+        self.template = ReceiptTemplate.get_or_404(request.view_args['template_id'])
+
+
+class ReceiptAreaRenderMixin(ReceiptAreaMixin):
+    def _render_template(self, tpl_name, **kwargs):
+        view_class = WPEventReceiptTemplates if self.object_type == 'event' else WPCategoryReceiptTemplates
+        return view_class.render_template(tpl_name, self.target, 'receipts', target=self.target, **kwargs)
+
+
+class TemplateListMixin(ReceiptAreaRenderMixin):
+    def _process(self):
+        inherited_templates = ReceiptTemplateDBSchema(many=True).dump(get_inherited_templates(self.target))
+        own_templates = ReceiptTemplateDBSchema(many=True).dump(self.target.receipt_templates)
+        return self._render_template(
+            'list.html',
+            inherited_templates=inherited_templates,
+            own_templates=own_templates,
+            target_locator=self.target.locator
+        )
+
+
+class AllTemplateMixin(ReceiptAreaMixin):
+    def _process(self, **data):
+        schema = ReceiptTemplateDBSchema(only=('id', 'title', 'custom_fields', 'default_filename'), many=True)
+        inherited_templates = schema.dump(get_inherited_templates(self.target))
+        own_templates = schema.dump(self.target.receipt_templates)
+        return jsonify(inherited_templates + own_templates)
+
+
+class RHListEventTemplates(TemplateListMixin, RHManageEventBase):
+    pass
+
+
+class RHListCategoryTemplates(TemplateListMixin, RHManageCategoryBase):
+    pass
+
+
+class RHAddTemplate(ReceiptAreaRenderMixin, RHAdminBase):
+    always_clonable = False
+
+    @use_kwargs(ReceiptTemplateAPISchema)
+    def _process(self, title, default_filename, html, css, yaml):
+        if yaml is None:
+            yaml = {'custom_fields': []}
+        template = ReceiptTemplate(title=title, html=html, css=css, custom_fields=yaml['custom_fields'],
+                                   default_filename=default_filename, **self.target_dict)
+        db.session.flush()
+        return jsonify(ReceiptTemplateDBSchema().dump(template))
+
+
+class DisplayTemplateMixin(ReceiptAreaRenderMixin):
+    def _process(self):
+        inherited_templates = ReceiptTemplateDBSchema(many=True).dump(get_inherited_templates(self.target))
+        own_templates = ReceiptTemplateDBSchema(many=True).dump(self.target.receipt_templates)
+        return self._render_template(
+            'list.html',
+            inherited_templates=inherited_templates,
+            own_templates=own_templates,
+            target_locator=self.target.locator
+        )
+
+
+class RHEventTemplate(DisplayTemplateMixin, RHManageEventBase):
+    pass
+
+
+class RHCategoryTemplate(DisplayTemplateMixin, RHManageCategoryBase):
+    pass
+
+
+class RHDeleteTemplate(ReceiptTemplateMixin, RHAdminBase):
+    def _process(self):
+        db.session.delete(self.template)
```

Comment:
What happens with already-generated receipts? Those must not be lost!
So maybe we'll have to do soft deletion here?

Reply:
Indeed, documents got deleted. Luckily changing it to soft deletion was pretty easy.

To update the local DB w/o losing all the templates:

```sql
alter table "indico"."receipt_templates" add column "is_deleted" boolean not null default false;
alter table "indico"."receipt_templates" alter column "is_deleted" drop default;
```